strawberryfields/apps/qchem/dynamics.py

```diff
@@ -250,3 +254,81 @@ def prob(samples: list, excited_state: list) -> float:
         raise ValueError("The excited state must not contain negative values")
 
     return samples.count(excited_state) / len(samples)
+
+
+def sample_tmsv(
+    r: list, t: float, Ul: np.ndarray, w: np.ndarray, n_samples: int, loss: float = 0.0,
+) -> list:
+    r"""Generate samples for simulating vibrational quantum dynamics with an input two-mode squeezed state.
+
+    Given N two-mode squeezing parameters and N-dim normal-to-local transformation matrix, Ul,
+    a GBS device with 2N modes is simulated. The evolution operator acts on only the first N
+    modes in the device. A sample is generated by measuring the number of photons in each of
+    the :math:`2N` modes.
+
+    **Example usage:**
+
+    >>> r = [0.2, 0.8]
+    >>> t = 10.0
+    >>> Ul = np.array([[0.707106781, -0.707106781],
+    >>>                [0.707106781, 0.707106781]])
+    >>> w = np.array([3914.92, 3787.59])
+    >>> n_samples = 5
+    >>> sample_tmsv(r, t, Ul, w, n_samples)
+    [[0, 0, 0, 0], [0, 1, 0, 1], [0, 2, 0, 2], [0, 0, 0, 0], [1, 1, 1, 1]]
+
+    Args:
+        r (list): list of two-mode squeezing parameters
+        t (float): time in femtoseconds
+        Ul (array): normal-to-local transformation matrix
+        w (array): normal mode frequencies :math:`\omega` in units of :math:`\mbox{cm}^{-1}`
+        n_samples (int): number of samples to be generated
+        loss (float): loss parameter denoting the fraction of lost photons
+
+    Returns:
+        list[list[int]]: a list of samples
+    """
+    if n_samples < 1:
+        raise ValueError("Number of samples must be at least one")
+
+    if t < 0:
+        raise ValueError("Time must be zero or positive")
+
+    if np.any(w <= 0):
+        raise ValueError("Vibrational frequencies must be larger than zero")
+
+    if np.any(np.iscomplex(Ul)):
+        raise ValueError("The normal mode to local mode transformation matrix must be real")
+
+    if not len(r) == len(Ul):
+        raise ValueError(
+            "Number of squeezing parameters and the number of modes in the normal-to-local"
+            " transformation matrix must be equal"
+        )
+
+    N = len(Ul)
+    op = evolution(N)
+
+    eng = sf.LocalEngine(backend="gaussian")
+    prog = sf.Program(2 * N)
+
+    # pylint: disable=expression-not-assigned,pointless-statement
+    with prog.context as q:
+
+        for i in range(N):
+            sf.ops.S2gate(r[i]) | (q[i], q[i + N])
+
+        op(t, Ul, w) | q[:N]
+
+        if loss:
+            for _q in q:
+                sf.ops.LossChannel(1 - loss) | _q
+
+        sf.ops.MeasureFock() | q
+
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", category=UserWarning, message="Cannot simulate non-")
+
+        s = eng.run(prog, shots=n_samples).samples
+
+    return s.tolist()
```

Comment:
This comment is true also for PR #423 : Why do we convert to list here?

Reply:
list of samples is a more elegant output. `eng.run(prog).samples` returns an array of samples with Gaussian backend. It returns even more things when fock backend is used. so for PR #423 , we have to further process the engine result. The user only sees the list of samples that are really easy to use further.

Reply:
@trbromley do you agree? I remember we had some issues in the apps layer when using lists instead of arrays

Reply:
We use `return s.tolist()` in `apps.sample`.

Reply:
Sorry, only just got around to replying on the other thread! https://github.com/XanaduAI/strawberryfields/pull/423/files#r447988839
TLDR: I think it's ok as it follows what we do in the rest of the apps layer, now including in the `qchem` module, but maybe we can change going forward.